### PR TITLE
feat(rbac): role overview

### PR DESCRIPTION
**Pull request by debsmita1**
Resolves:
https://github.com/janus-idp/backstage-plugins/issues/854

**GIF/Screenshots:**
https://github.com/janus-idp/backstage-plugins/assets/22490998/8583257f-eaae-4503-a781-0fa57c915499

Overview Page
Screenshot 2023-11-28 at 9 23 46 PM

Card with Error
Screenshot 2023-12-01 at 12 26 14 AM

**Test setup:**

Follow the steps mentioned in the PR https://github.com/janus-idp/backstage-plugins/pull/895 and the [documentation](https://github.com/janus-idp/backstage-plugins/tree/main/plugins/rbac) to install the RBAC UI plugin

Add 
```app-config.yaml
catalog:
  locations:
     ...
     - type: URL
        target: https://github.com/backstage/backstage/blob/master/packages/catalog-model/examples/acme-corp.yaml
        rules:
          - allow: [User, Group]
```

### CODE REVIEW

**Comment by parvathyvr** [comment]
@debsmita1 the changes needed are:
1. Members table name to 'Users and groups'.
2. The number of 'Permission policies' we should show is the number of policies- Eg: if there is a single row and there are 'Read,Update,Delete' those are 3 policies.

File: plugins/rbac/package.json [hunk 1]
```diff
@@ -25,16 +25,17 @@
...
+    "@janus-idp/backstage-plugin-rbac-common": "1.1.0",
     "@material-ui/core": "^4.9.13",
     "@material-ui/icons": "^4.11.3",
     "@material-ui/lab": "^4.0.0-alpha.45",
-    "react-use": "^17.4.0",
-    "@backstage/plugin-permission-react": "^0.4.16",
-    "@janus-idp/backstage-plugin-rbac-common": "1.1.0",
-    "@mui/icons-material": "5.14.11"
+    "@mui/icons-material": "5.14.11",
+    "react-use": "^17.4.0"
```
Comment (divyanshiGupta) [inline plugins/rbac/package.json:38]: @debsmita1 what is being updated in this file?

File: plugins/rbac/src/components/AboutCard.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  const classes = useStyles();
+  const cardClass = classes.gridItemCard;
+  const cardContentClass = classes.gridItemCardContent;
+
+  return (
+    <Card className={cardClass}>
+      <CardHeader
+        title="About"
+        action={
+          <IconButton aria-label="Refresh" title="Refresh" onClick={() => {}}>
```
Comment (divyanshiGupta) [inline plugins/rbac/src/components/AboutCard.tsx:49]: @debsmita1 is it like a placeholder icon without any cta for now?

**Review by divyanshiGupta — CHANGES_REQUESTED** [review]

**Comment by ShiranHi** [comment]
Hey @debsmita1, thanks! Noticed a few things:

1. The "role: default/guests" header looks a bit off – thinking it should have a background like the other catalog pages, also the tabs here seem to be floating
2. The about card needs to be wider, keeping it to a single row of content. "Modified by" and "Last modified" should share the row with the "Description" text (see [design](https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=977%3A7407&mode=design&t=uATaSCYoWnVSzNaB-1))
3. The edit button is missing in all cards
4. There's this empty space above each table, right before the column titles. Is that the table header? Can we remove it?

@parvathyvr FYI

**Comment by nirfarkas** [comment]
Another small nit:
Cards should be aligned left to the header text (for example at the beginning of the video we can see that the cards are aligned to the left side of the "Administration" header, while at 00:03 we can see that they jumped to the right and are not aligned with the header.
image
image

**Comment by debsmita1** [comment]
> Hey @debsmita1, thanks! Noticed a few things:
> 
> 1. The "role: default/guests" header looks a bit off – thinking it should have a background like the other catalog pages, also the tabs here seem to be floating
> 2. The about card needs to be wider, keeping it to a single row of content. "Modified by" and "Last modified" should share the row with the "Description" text (see [design](https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=977%3A7407&mode=design&t=uATaSCYoWnVSzNaB-1))
> 3. The edit button is missing in all cards
> 4. There's this empty space above each table, right before the column titles. Is that the table header? Can we remove it?
> 
> @parvathyvr FYI

Updated
Screenshot 2023-11-27 at 8 34 45 PM

cc @nirfarkas

File: plugins/rbac/package.json (hunk 1 above)
Comment (debsmita1) [reply plugins/rbac/package.json:40]: some libraries have been added

**Comment by parvathyvr** [comment]
@debsmita1 Thanks for the updated screen. Few things missing:
1. For the Permission policies table-the number shown should be the number of policies selected.
2. For the Users and groups table, for type user we could keep it blank as in the [design](https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=977-7407&mode=design&t=IROic36Q2uiUliEC-4). Because i don’t see a members card in the User overview page.

**Comment by debsmita1** [comment]
> @debsmita1 Thanks for the updated screen. Few things missing:
> 
> 1. For the Permission policies table-the number shown should be the number of policies selected.
> 2. For the Users and groups table, for type user we could keep it blank as in the [design](https://www.figma.com/file/E6C9X3OSVzlWkW7ZsbxN7s/Administration?type=design&node-id=977-7407&mode=design&t=IROic36Q2uiUliEC-4). Because i don’t see a members card in the User overview page.

Screenshot 2023-11-27 at 11 40 19 PM

**Comment by divyanshiGupta** [comment]
@parvathyvr I was wondering if we should have a search bar in the members and permissions tables? Wdyt?

**Comment by ShiranHi** [comment]
> * The edit button is missing in all cards
I think this is still missing

**Comment by debsmita1** [comment]
> Details

@ShiranHi We haven't implemented Edit yet. There's a separate story to handle it

**Comment by debsmita1** [comment]
> > * The edit button is missing in all cards
> 
> I think this is still missing

@ShiranHi We haven't implemented Edit yet. There's a separate story to handle it

**Comment by parvathyvr** [comment]
> @parvathyvr I was wondering if we should have a search bar in the members and permissions tables? Wdyt?

@divyanshiGupta it's a good idea! I didn't add it since i haven't noticed search in the cards of Overview page! But if we can incorporate search into the table,it's a good addition. CC @debsmita1

**Comment by debsmita1** [comment]
> > @parvathyvr I was wondering if we should have a search bar in the members and permissions tables? Wdyt?
> 
> @divyanshiGupta it's a good idea! I didn't add it since i haven't noticed search in the cards of Overview page! But if we can incorporate search into the table,it's a good addition. CC @debsmita1

Done

Screenshot 2023-11-28 at 9 23 46 PM

File: plugins/rbac/src/hooks/useMembers.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    return await rbacApi.getRole(roleName);
+  });
+
+  const { loading: membersLoading, value: members } = useAsync(async () => {
+    return await rbacApi.getMembers();
+  });
+
+  const data = React.useMemo(
+    () =>
+      role?.[0].memberReferences.reduce((acc: MembersData[], ref) => {
```
Comment (invincibleJai) [inline plugins/rbac/src/hooks/useMembers.ts:27]: why do we need `0` here, you are quering based on the role name so should be just one role right? Is it  due to the response

File: plugins/rbac/src/hooks/usePermissionPolicies.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+export const usePermissionPolicies = (entityReference: string) => {
+  const rbacApi = useApi(rbacApiRef);
+  const {
+    loading: policiesLoading,
+    value: policies,
+    retry,
+  } = useAsyncRetry(async () => {
+    return await rbacApi.getPolicies();
+  });
+  const { value: permissionPolicies } = useAsync(async () => {
```
Comment (invincibleJai) [inline plugins/rbac/src/hooks/usePermissionPolicies.ts:18]: would be good to check on loading here and in general good to check on error cased as well

**Review by invincibleJai — COMMENTED** [review]
Thanks @debsmita1 looks good overall, good to add unit tests for hooks as well

File: plugins/rbac/src/hooks/useMembers.ts (hunk 3 above)
Comment (debsmita1) [reply plugins/rbac/src/hooks/useMembers.ts:27]: yes, the response is an array with one role

File: plugins/rbac/src/hooks/usePermissionPolicies.ts (hunk 4 above)
Comment (debsmita1) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts:18]: How do we want to show the user if there is failure in API call ?
Note: Each card makes different API calls cc @parvathyvr @ShiranHi

Comment (parvathyvr) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: > How do we want to show the user if there is failure in API call ? Note: Each card makes different API calls cc @parvathyvr @ShiranHi

@debsmita1 Does it make sense to use the [error banner](https://www.figma.com/file/wlH5SQTQiIMHp5T0t7P4aU/Backstage-Design-System-(Community)?type=design&node-id=137-772&mode=design&t=is0zKoWImNDw28Ap-4) in backstage design system for this scenario?

Comment (debsmita1) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: Screenshot 2023-11-29 at 10 29 49 PM

Comment (invincibleJai) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: @debsmita1 I know this is an edge case and won't happen ideally but should we show the error above the table under each card rather than inside of the table?

@parvathyvr WDYT?

Comment (parvathyvr) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: @invincibleJai i thought we could replace the 'No records to display' with the error message as @debsmita1 has implemented-But I am okay with the error being above the empty table as well- If that is a better option!
CC @ShiranHi do we have a preference/guideline about this?

Comment (invincibleJai) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: okay nice if you have considered this 👍

Comment (debsmita1) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: Updated

Screenshot 2023-11-30 at 1 52 50 PM

Comment (ShiranHi) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: @debsmita1 it looks good! Can we add a bit of a 16px space between the error message and the table ?
Also, it'll be great if a word will not break into two lines "grou-ps" and "poli-cies" like it is right now. 
Thanks for all the changes!

Comment (debsmita1) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: Screenshot 2023-12-01 at 12 26 14 AM

Comment (parvathyvr) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: A small note- @debsmita1 we are showing 'No members found' and 'No permissions found' on the empty table.
But ideally since the table heading are 'Users and groups' and 'Permission policies' i think it should be 'No users and groups found' and 'No permission policies found'.
But i also wonder if it's easier to show the generic message' No records found'.

Comment (nirfarkas) [reply plugins/rbac/src/hooks/usePermissionPolicies.ts]: > The word breaks on small screens, what do you suggest here ?

I would suggest changing to 1 column per table to avoid text breaks as well as table misalignment.
Another reason to avoid 2 column composition is to reduce the cognitive load and showing two elements next to each other is usually for comparing them (which is not the case here).
So,
TL;DR
Moving to 1 column layout will help with:

- Get rid of the text breaks
- Allow easier data consumption with no distractions

**Comment by debsmita1** [comment]
> @debsmita1 I see a mismatch in the number of users/groups and the policies in list page vs overview page
> 
> image image
> the reason for the user is, that the user is not registered however if we get the user in API we should show it irrespective of whether it's registered or not.
> 
> and the reason for policies in in list we are showing count of all the policies irrespective of effect allow or deny but in overview we only list allow once hence the mismatch @parvathyvr Let us know your thoughts here

Addressed this, now listing all users received in reponse, if a user doesn't exist simply showing "Entity not found" when the user is clicked.

Also, not listing a role if it has no users/ policies associated with it

**Review by invincibleJai — APPROVED** [review]
LGTM

Thanks @debsmita1

**Review by parvathyvr — APPROVED** [review]
@debsmita1 LGTM! Thank you.
